client/blocks/signup-form/passwordless.jsx

```diff
@@ -0,0 +1,260 @@
+import config from '@automattic/calypso-config';
+import { Button } from '@automattic/components';
+import emailValidator from 'email-validator';
+import { localize } from 'i18n-calypso';
+import PropTypes from 'prop-types';
+import { Component } from 'react';
+import { connect } from 'react-redux';
+import FormLabel from 'calypso/components/forms/form-label';
+import FormTextInput from 'calypso/components/forms/form-text-input';
+import LoggedOutForm from 'calypso/components/logged-out-form';
+import LoggedOutFormFooter from 'calypso/components/logged-out-form/footer';
+import Notice from 'calypso/components/notice';
+import { recordRegistration } from 'calypso/lib/analytics/signup';
+import { getLocaleSlug } from 'calypso/lib/i18n-utils';
+import wpcom from 'calypso/lib/wp';
+import ValidationFieldset from 'calypso/signup/validation-fieldset';
+import { recordTracksEvent } from 'calypso/state/analytics/actions';
+import { saveSignupStep, submitSignupStep } from 'calypso/state/signup/progress/actions';
+
+class PasswordlessSignupForm extends Component {
+	static propTypes = {
+		locale: PropTypes.string,
+	};
+
+	static defaultProps = {
+		locale: 'en',
+	};
+
+	state = {
+		isSubmitting: false,
+		email: this.props.step && this.props.step.form ? this.props.step.form.email : '',
+		errorMessages: null,
+	};
+
+	submitTracksEvent = ( isSuccessful, props ) => {
+		const tracksEventName = isSuccessful
+			? 'calypso_signup_actions_onboarding_passwordless_login_success'
+			: 'calypso_signup_actions_onboarding_passwordless_login_error';
+		this.props.recordTracksEvent( tracksEventName, {
+			...props,
+		} );
+	};
+
+	onFormSubmit = async ( event ) => {
+		event.preventDefault();
+
+		if ( ! this.state.email || ! emailValidator.validate( this.state.email ) ) {
```

Comment:
Updating the state of a component based on the current value of `this.state` may potentially lead to inconsistent component state.

This is also documented in the [React docs](https://reactjs.org/docs/react-component.html#setstate):

> `setState()` does not always immediately update the component. It may batch or defer the update until later. This makes reading `this.state` right after calling `setState()` a potential pitfall. Instead, use `componentDidUpdate` or a `setState` callback (`setState(updater, callback)`)

The easiest way to fix this is to just change `setState()` to the callback form, where it will always properly access the current `state.email` value.

Reply:
Thanks @tyxla, I've run into state timing issues with useEffect, but I'm not as familiar with the class-based nuances. Quick question about this - wouldn't it be okay how it's used here since it reads the state first _then_ sets state based on that value?

Reply:
I've seen problems happening in complex components when we access it before we update the state, too. At the same time, it's very hard to reproduce those issues, so it is mostly a theoretical issue. I'll leave it up to you, but I'd suggest fixing it to ensure `this.state` has the expected up-to-date value in that condition.